core/src/main/scala/kafka/log/Log.scala

```diff
@@ -582,6 +582,64 @@ class Log(val dir: File,
     }
   }
 
+  private class SizeBasedRetentionCheck(private var diff: Long) {
+    def shouldDelete(segment: LogSegment) = {
+      if(diff - segment.size >= 0) {
+        diff -= segment.size
+        true
+      } else {
+        false
+      }
+    }
+  }
+
+  /**
+    * Delete any log segments that have either expired due to time based retention
+    * or because the log size is > retentionSize
+    */
+  def deleteOldSegments(): Int = {
+    if (!config.delete) return 0
+
+    var deleted = 0
+    if (config.retentionMs >= 0) {
+      val startMs = time.milliseconds
+      deleted += deleteOldSegments(startMs - _.lastModified > config.retentionMs)
+    }
+    if (config.retentionSize >= 0 && size > config.retentionSize) {
+      deleted += deleteOldSegments(new SizeBasedRetentionCheck(size - config.retentionSize).shouldDelete)
+    }
+    deleted
+  }
+
+  private def hasOldSegments(predicate: LogSegment => Boolean): Boolean = {
+    lock synchronized {
+      val lastEntry = segments.lastEntry
+      if (lastEntry == null) false
+      else logSegments.exists(s => predicate(s) && (s.baseOffset != lastEntry.getValue.baseOffset || s.size > 0))
+    }
+  }
+
+
+
+  def shouldDeleteSegments: Boolean = {
+    if (!config.delete) return false
+
+    def retentionMsExpired():Boolean = {
```

Comment:
Could we share this code with those in line 604 somehow? Ditto for retentionSizeBreached().